### PR TITLE
feat: merkle verification and changes implemented

### PR DESCRIPTION
**Pull request by SamAg19**
## Problem statement:
The current implementation requires the signed Oracle block to be sent to the result manager to update collection values. This is not scalable and updating will become expensive as we add more collections to the Razor Oracle.

## Solution:
We will be shifting to a Merkle tree implementation for updation in which the Merkle leaves will hold the result of each collection and the clients will receive signed result, Merkle root, and Merkle proof along with the signature for each collection depending on the collection data they require

To update the values of each collection, clients will need to query the Merkle tree for each collection off-chain which will be exposed through an API, and send the data received through their contracts to update the values. The clients will only need to update the collections they require for their application.

The functions exposed to clients will be:

- `fetchResult()`:- this function is responsible for updating the collection data and receiving the latest collection result. The result will only be updated if the `lastUpdatedTimestamp` for a particular collection in the contracts precedes the one that is being sent over, else the result won't be updated and the client will receive the stored result in the contracts
- `getResult()`:- this is a view function where you will just receive the collection data stored in the contracts. It does not deal with updation and the results retrieved may be stale
- `validateResult()`: this is a view function that allows people to validate whether the data they are receiving from the API is valid or not

### CODE REVIEW

File: contracts/Forwarder.sol [hunk 1]
```diff
@@ -36,40 +38,114 @@ contract Forwarder is AccessControlEnumerable, Pausable {
...
     }
 
-    /// @notice get result by collection name
-    function getResult(bytes32 collectionName)
+    /**
+     * @notice Updates the result based on the provided data and returns the latest result
+     * @param data bytes data required to update the result
+     * @return result of the collection, its power and timestamp
+     */
+    function getResult(bytes calldata data)
```
Comment (yohanelly95) [inline contracts/Forwarder.sol:85]: should we rename this function to `getLatestResult` or `updateAndGetResult`, so that its implicit that this fn call would update and return the latest result always. 

As opposed to `getResult(bytes32 name)` which would return the result even if its not updated recently or stale. (this way might cost less gas, but stale results possible) @SamAg19 @SkandaBhat

File: contracts/Forwarder.sol [hunk 2]
```diff
@@ -36,40 +38,114 @@ contract Forwarder is AccessControlEnumerable, Pausable {
...
+     */
+    function validateResult(bytes calldata data)
+        external
+        view
+        whenNotPaused
+        onlyRole(TRANSPARENT_FORWARDER_ROLE)
+        returns (bool)
+    {
+        require(
+            validateSelector != bytes4(0),
```
Comment (yohanelly95) [inline contracts/Forwarder.sol:143]: we can add a modifier for this check, and maybe we can migrate to error types. Use modifier wherever required.

File: contracts/ResultManager.sol [hunk 3]
```diff
@@ -3,6 +3,7 @@ pragma solidity ^0.8.0;
 
 import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
 import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
 
 contract ResultManager is AccessControlEnumerable {
     struct Block {
```
Comment (yohanelly95) [inline contracts/ResultManager.sol:9]: remove this struct, this is being used as Value struct already

**Review by yohanelly95 — CHANGES_REQUESTED** [review]

File: contracts/Forwarder.sol (hunk 2 above)
Comment (SamAg19) [reply contracts/Forwarder.sol]: https://github.com/razor-network/bridge-proxy/pull/59/commits/00222f0d6f0d9912d66935ac1fd7dde93da0f06b

added modifier

File: contracts/ResultManager.sol (hunk 3 above)
Comment (SamAg19) [reply contracts/ResultManager.sol]: removed https://github.com/razor-network/bridge-proxy/pull/59/commits/00222f0d6f0d9912d66935ac1fd7dde93da0f06b

File: contracts/Forwarder.sol (hunk 1 above)
Comment (SamAg19) [reply contracts/Forwarder.sol]: Renamed to `getLatestResult` for now

File: contracts/Forwarder.sol (hunk 2 above)
Comment (SamAg19) [reply contracts/Forwarder.sol]: https://github.com/razor-network/bridge-proxy/pull/59/commits/857e9caac099e33c24e6fc54bb698f1e385f2821
Added custom error types

File: contracts/TransparentForwarder.sol [hunk 4]
```diff
@@ -41,7 +43,8 @@ contract TransparentForwarder is AccessControlEnumerable {
         bool isWhitelisted = staking.isWhitelisted{value: msg.value}(
             msg.sender
         );
-        require(isWhitelisted, "Not whitelisted");
```
Comment (yohanelly95) [inline contracts/TransparentForwarder.sol:44]: can revert this change, since we should avoid having to redeploy transparent forwarder for existing testnets @SamAg19

**Review by yohanelly95 — APPROVED** [review]

File: merkle-tree.json
Comment (rajkharvar) [inline merkle-tree.json:1]: We can delete this file. Not required

Comment (SamAg19) [reply merkle-tree.json:1]: removed https://github.com/razor-network/bridge-proxy/pull/59/commits/45493e6645c4faf66d29b56a095fcd59aacd9073

**Review by yohanelly95 — APPROVED** [review]

**Review by rajkharvar — APPROVED** [review]

File: contracts/ResultManager.sol [hunk 5]
```diff
@@ -3,84 +3,123 @@ pragma solidity ^0.8.0;
...
+     * @return result of the collection, its power and timestamp
+    */
+    function updateResult(
+        bytes32 merkleRoot,
+        bytes32[] memory proof,
+        Value memory result,
+        bytes memory signature
+    ) external onlyRole(FORWARDER_ROLE) returns (uint256, int8, uint256) {
+        if (result.lastUpdatedTimestamp > _collectionResults[result.name].lastUpdatedTimestamp) {
+            bytes memory resultBytes = abi.encode(result);
```
Comment (hrishikeshio) [inline contracts/ResultManager.sol:67]: not required

File: contracts/ResultManager.sol [hunk 6]
```diff
@@ -3,84 +3,123 @@ pragma solidity ^0.8.0;
...
+            bytes32 messageHash = keccak256(
+                abi.encodePacked(merkleRoot, resultBytes)
```
Comment (hrishikeshio) [inline contracts/ResultManager.sol:69]: just sign merkle root

File: contracts/ResultManager.sol [hunk 7]
```diff
@@ -3,84 +3,123 @@ pragma solidity ^0.8.0;
...
+            );
+            if(
+                ECDSA.recover(
+                    ECDSA.toEthSignedMessageHash(messageHash),
+                    signature
+                ) != signerAddress) revert InvalidSignature();
+
+            bytes32 leaf = keccak256(
+                bytes.concat(keccak256(abi.encode(resultBytes)))
```
Comment (hrishikeshio) [inline contracts/ResultManager.sol:78]: result not resultbytes

File: contracts/ResultManager.sol [hunk 8]
```diff
@@ -3,84 +3,123 @@ pragma solidity ^0.8.0;
...
+            );
+            if(
+                !MerkleProof.verify(proof, merkleRoot, leaf)
+            ) revert InvalidMerkleProof();
+            _collectionResults[result.name] = result;
+
+            emit ResultUpdated(result);
+        }
+
+        return _getResult(result.name);
```
Comment (hrishikeshio) [inline contracts/ResultManager.sol:88]: In future optimization, add multiproof support.

File: contracts/ResultManager.sol [hunk 9]
```diff
@@ -3,84 +3,123 @@ pragma solidity ^0.8.0;
...
-        }
-        lastUpdatedTimestamp = timestamp;
-        latestEpoch = epoch;
+        bytes32 leaf = keccak256(
+            bytes.concat(keccak256(abi.encode(resultBytes)))
+        );
+        if (!MerkleProof.verify(proof, merkleRoot, leaf)) return false;
 
-        emit BlockReceived(messageBlock);
+        return true;
```
Comment (hrishikeshio) [inline contracts/ResultManager.sol:122]: code repetition.

File: contracts/interface/IDelegator.sol [hunk 10]
```diff
@@ -3,80 +3,31 @@ pragma solidity ^0.8.0;
 
 interface IDelegator {
```
Comment (hrishikeshio) [inline contracts/interface/IDelegator.sol:4]: change name to iforwarder

**Review by hrishikeshio — COMMENTED** [review]
changes requested
